virtcontainers/sandbox.go

```diff
@@ -931,6 +934,122 @@ func (s *Sandbox) removeNetwork() error {
+}
+
+func (s *Sandbox) generateNetInfo(inf grpc.Interface) (NetworkInfo, error) {
+	hw, err := net.ParseMAC(inf.HwAddr)
+	if err != nil {
+		return NetworkInfo{}, err
+	}
+
+	var addrs []netlink.Addr
+	for _, addr := range inf.IPAddresses {
+		netlinkAddrStr := fmt.Sprintf("%s/%s", addr.Address, addr.Mask)
+		netlinkAddr, err := netlink.ParseAddr(netlinkAddrStr)
+		if err != nil {
+			return NetworkInfo{}, fmt.Errorf("could not parse %q: %v", netlinkAddrStr, err)
+		}
+
+		addrs = append(addrs, *netlinkAddr)
+	}
+
+	return NetworkInfo{
+		Iface: NetlinkIface{
+			LinkAttrs: netlink.LinkAttrs{
+				Name:         inf.Name,
+				HardwareAddr: hw,
+				MTU:          int(inf.Mtu),
+			},
+			Type: "",
+		},
+		Addrs: addrs,
+	}, nil
+}
+
+func (s *Sandbox) generateInf(netInfo NetworkInfo) (grpc.Interface, error) {
+	iface := netInfo.Iface
+	hw := iface.HardwareAddr.String()
+
+	return grpc.Interface{
+		Device:      iface.Name,
+		Name:        iface.Name,
+		IPAddresses: generateIPAddr(netInfo.Addrs, s.Logger()),
+		Mtu:         uint64(iface.MTU),
+		HwAddr:      hw,
+	}, nil
+}
+
+// AddNetwork adds new nic to the sandbox.
+func (s *Sandbox) AddNetwork(inf grpc.Interface) error {
+	netInfo, err := s.generateNetInfo(inf)
+	if err != nil {
+		return err
+	}
+
+	endpoint, err := createVirtualNetworkEndpoint(len(s.networkNS.Endpoints), netInfo.Iface.Name, s.config.NetworkConfig.InterworkingModel)
+	if err != nil {
+		return err
+	}
+	endpoint.SetProperties(netInfo)
+	if err := endpoint.HotAttach(s.hypervisor); err != nil {
+		return err
+	}
+
+	// Update the sandbox storage
+	s.networkNS.Endpoints = append(s.networkNS.Endpoints, endpoint)
+	if err := s.storage.storeSandboxNetwork(s.id, s.networkNS); err != nil {
+		return err
+	}
+
+	// Add network for vm
+	if err := s.agent.updateNetwork(endpoint); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ListNetwork lists all nics and their configurations in the sandbox.
+func (s *Sandbox) ListNetwork() ([]grpc.Interface, []*grpc.Route, error) {
+	return s.networkNS.Interfaces, s.networkNS.Routes, nil
+}
+
+// UpdateNetwork updates the configuration of an existing network interface.
+func (s *Sandbox) UpdateNetwork(interfaces []grpc.Interface) error {
+	if err := s.deleteNetworks(); err != nil {
+		return err
+	}
+	for _, inf := range interfaces {
+		if err := s.AddNetwork(inf); err != nil {
+			return err
+		}
+	}
+	s.networkNS.Interfaces = interfaces
+	return s.storage.storeSandboxNetwork(s.id, s.networkNS)
+}
+
+// deleteNetworks deletes all existing network interfaces.
+func (s *Sandbox) deleteNetworks() error {
+	for i := len(s.networkNS.Endpoints) - 1; i >= 0; i-- {
+		endpoint := s.networkNS.Endpoints[i]
+		if err := endpoint.HotDetach(s.hypervisor, s.networkNS.NetNsCreated, s.networkNS.NetNsPath); err != nil {
+			return err
+		}
+		s.networkNS.Endpoints = s.networkNS.Endpoints[:i]
+		if err := s.storage.storeSandboxNetwork(s.id, s.networkNS); err != nil {
```

Comment:
Each add network and each delete network would save the network config once. From metadata persistency point of view, can we simplify it to just do whatever configuration the API needs, and then save the persistent data once before returning? This would save many IO operations and speed up network operations.